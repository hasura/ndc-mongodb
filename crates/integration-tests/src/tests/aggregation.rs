use insta::assert_yaml_snapshot;
use serde_json::json;

use crate::graphql_query;

#[tokio::test]
async fn runs_aggregation_over_top_level_fields() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query($albumId: Int!) {
                  track(order_by: { id: Asc }, where: { albumId: { _eq: $albumId } }) {
                    milliseconds
                    unitPrice
                  }
                  trackAggregate(
                    filter_input: { order_by: { id: Asc }, where: { albumId: { _eq: $albumId } } }
                  ) {
                    _count
                    milliseconds {
                      _avg
                      _max
                      _min
                      _sum
                    }
                    unitPrice {
                      _count
                      _count_distinct
                    }
                  }
                }
            "#
        )
        .variables(json!({ "albumId": 9 }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_extended_json_representing_mixture_of_numeric_types() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This doesn't affect native queries that don't use the $documents stage.
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      _avg
                      _count
                      _max
                      _min
                      _sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "decimal|double|int|long" }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_mixture_of_numeric_and_null_values() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This doesn't affect native queries that don't use the $documents stage.
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      _avg
                      _count
                      _max
                      _min
                      _sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "double|null" }))
        .run()
        .await?
    );
    Ok(())
}
