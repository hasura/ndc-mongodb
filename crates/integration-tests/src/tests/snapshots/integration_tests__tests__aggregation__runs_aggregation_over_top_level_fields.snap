---
source: crates/integration-tests/src/tests/aggregation.rs
expression: "graphql_query(r#\"\n                query($albumId: Int!) {\n                  track(order_by: { id: Asc }, where: { albumId: { _eq: $albumId } }) {\n                    milliseconds\n                    unitPrice\n                  }\n                  trackAggregate(\n                    filter_input: { order_by: { id: Asc }, where: { albumId: { _eq: $albumId } } }\n                  ) {\n                    _count\n                    milliseconds {\n                      avg\n                      max\n                      min\n                      sum\n                    }\n                    unitPrice {\n                      _count\n                      _count_distinct\n                    }\n                  }\n                }\n            \"#).variables(json!({\n    \"albumId\": 9\n})).run().await?"
---
data:
  track:
    - milliseconds: 221701
      unitPrice: "0.99"
    - milliseconds: 436453
      unitPrice: "0.99"
    - milliseconds: 374543
      unitPrice: "0.99"
    - milliseconds: 322925
      unitPrice: "0.99"
    - milliseconds: 288208
      unitPrice: "0.99"
    - milliseconds: 308035
      unitPrice: "0.99"
    - milliseconds: 369345
      unitPrice: "0.99"
    - milliseconds: 350197
      unitPrice: "0.99"
  trackAggregate:
    _count: 8
    milliseconds:
      avg: 333925.875
      max: 436453
      min: 221701
      sum: 2671407
    unitPrice:
      _count: 8
      _count_distinct: 1
errors: ~
